crates/ghost_actor/src/ghost_channel.rs

```diff
@@ -0,0 +1,271 @@
+{
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "GhostMessage {{ .. }}")
+    }
+}
+
+impl<RequestToSelf, RequestToOther, RequestToSelfResponse, Error>
+    GhostMessage<RequestToSelf, RequestToOther, RequestToSelfResponse, Error>
+{
+    fn new(
+        request_id: Option<RequestId>,
+        payload: RequestToSelf,
+        sender: crossbeam_channel::Sender<
+            GhostChannelMessage<RequestToOther, RequestToSelfResponse, Error>,
+        >,
+    ) -> Self {
+        Self {
+            request_id,
+            payload: Some(payload),
+            sender,
+        }
+    }
+
+    pub fn take_payload(&mut self) -> Option<RequestToSelf> {
+        std::mem::replace(&mut self.payload, None)
+    }
+
+    pub fn respond(self, payload: Result<RequestToSelfResponse, Error>) {
+        if let Some(request_id) = &self.request_id {
+            self.sender
+                .send(GhostChannelMessage::Response {
+                    request_id: request_id.clone(),
+                    payload,
+                })
+                .expect("should send");
+        }
+    }
+}
+
+pub struct GhostChannel<
+    RequestToOther,
+    RequestToOtherResponse,
+    RequestToSelf,
+    RequestToSelfResponse,
+    Error,
+> {
+    sender: crossbeam_channel::Sender<
+        GhostChannelMessage<RequestToOther, RequestToSelfResponse, Error>,
+    >,
+    receiver: crossbeam_channel::Receiver<
+        GhostChannelMessage<RequestToSelf, RequestToOtherResponse, Error>,
+    >,
+}
+
+impl<RequestToOther, RequestToOtherResponse, RequestToSelf, RequestToSelfResponse, Error>
+    GhostChannel<
+        RequestToOther,
+        RequestToOtherResponse,
+        RequestToSelf,
+        RequestToSelfResponse,
+        Error,
+    >
+{
+    fn new(
+        sender: crossbeam_channel::Sender<
+            GhostChannelMessage<RequestToOther, RequestToSelfResponse, Error>,
+        >,
+        receiver: crossbeam_channel::Receiver<
+            GhostChannelMessage<RequestToSelf, RequestToOtherResponse, Error>,
+        >,
+    ) -> Self {
+        Self { sender, receiver }
+    }
+
+    pub fn as_context_channel<Context>(
+        self,
+    ) -> GhostContextChannel<
+        Context,
+        RequestToOther,
+        RequestToOtherResponse,
+        RequestToSelf,
+        RequestToSelfResponse,
+        Error,
+    > {
+        GhostContextChannel::new(self.sender, self.receiver)
+    }
+}
+
+pub struct GhostContextChannel<
+    Context,
+    RequestToOther,
+    RequestToOtherResponse,
+    RequestToSelf,
+    RequestToSelfResponse,
+    Error,
+> {
+    sender: crossbeam_channel::Sender<
+        GhostChannelMessage<RequestToOther, RequestToSelfResponse, Error>,
+    >,
+    receiver: crossbeam_channel::Receiver<
+        GhostChannelMessage<RequestToSelf, RequestToOtherResponse, Error>,
+    >,
+    await_request_to_other_response: GhostTracker<Context, RequestToOtherResponse, Error>,
```

Comment:
this name confused me.  shouldn't we either have `tracker` or `callbacks` in the name?